crates/miden-objects/src/asset/mod.rs

```diff
@@ -198,13 +198,16 @@ impl TryFrom<Word> for Asset {
     fn try_from(value: Word) -> Result<Self, Self::Error> {
         // Return an error if element 3 is not a valid account ID prefix, which cannot be checked by
         // is_not_a_non_fungible_asset.
-        AccountIdPrefix::try_from(value[3])
+        let aip = AccountIdPrefix::try_from(value[3])
             .map_err(|err| AssetError::InvalidFaucetAccountId(Box::from(err)))?;
-
-        if is_not_a_non_fungible_asset(value) {
-            FungibleAsset::try_from(value).map(Asset::from)
-        } else {
-            NonFungibleAsset::try_from(value).map(Asset::from)
+        match aip.account_type() {
+            AccountType::FungibleFaucet => FungibleAsset::try_from(value).map(Asset::from),
+            AccountType::NonFungibleFaucet => NonFungibleAsset::try_from(value).map(Asset::from),
+            _ => {
+                let account_id = AccountId::try_from([value[3], value[2]])
+                    .map_err(|err| AssetError::InvalidFaucetAccountId(Box::new(err)))?;
+                Err(AssetError::FungibleFaucetIdTypeMismatch(account_id))
+            },
```

Comment:
This is a nice idea to improve this code, but a serialized non-fungible asset never includes the account ID suffix, only the prefix. So `AccountId::try_from` would always fail and I think we can just unconditionally return something like 
```rust
AssetError::InvalidFaucetAccountId(Box::from(
    format!("account ID prefix in asset contained a non-faucet ID: {aip}"
))
```

Reply:
Thank you! That's a very good catch